firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java

```diff
@@ -465,15 +465,15 @@ private void validateOrderByFieldMatchesInequality(
   }
 
   private void validateNewFilter(Filter filter) {
-    if (filter instanceof RelationFilter) {
-      Operator filterOp = ((RelationFilter) filter).getOperator();
+    if (filter instanceof FieldFilter) {
+      Operator filterOp = ((FieldFilter) filter).getOperator();
       List<Operator> arrayOps = Arrays.asList(Operator.ARRAY_CONTAINS, Operator.ARRAY_CONTAINS_ANY);
       List<Operator> disjunctiveOps = Arrays.asList(Operator.ARRAY_CONTAINS_ANY, Operator.IN);
       boolean isArrayOp = arrayOps.contains(filterOp);
       boolean isDisjunctiveOp = disjunctiveOps.contains(filterOp);
 
-      RelationFilter relationFilter = (RelationFilter) filter;
-      if (relationFilter.isInequality()) {
+      FieldFilter fieldFilter = (FieldFilter) filter;
```

Comment:
You could push this cast up above the declaration of `filterOp` to avoid casting twice.